client.js

```diff
@@ -20,38 +20,96 @@ const crypto = require("crypto");
 const objectOps = require("lib/objectOps.js");
 const fileOps = require("lib/fileOps");
 const stringUtils = require("lib/stringUtils.js");
-const configManager = require("lib/manager/configManager.js");
 const pluginManager = require("lib/manager/pluginManager");
+const modManager = require("lib/manager/modManager");
 const hashFile = require('lib/hash').hashFile;
 
+// Uhm...
+var global = {};
 
-// argument parsing
-const args = require('minimist')(process.argv.slice(2));
+/**
+ * Keeps track of the runtime parameters of an instance
+ */
+class Instance {
+	constructor(dir, name) {
+		this._name = name;
+		this._dir = dir;
+	}
 
-// require config.json
-var config = require(args.config || './config');
-var global = {};
+	/**
+	 * Name of the instance
+	 *
+	 * This should not be used for filesystem paths.  See .path() for that.
+	 */
+	get name() {
+		return this._name;
+	}
 
-const instance = process.argv[3];
-const instancedirectory = config.instanceDirectory + '/' + instance;
-const command = process.argv[2];
+	/**
+	 * Return path in instance
+	 *
+	 * Creates a path using path.join with the given parts that's relative to
+	 * the directory of the instance.  For example instance.path("mods")
+	 * returns a path to the mods directory of the instance.  If no parts are
+	 * given it returns a path to the directory of the instance.
+	 */
+	path(...parts) {
+		return path.join(this._dir, ...parts);
+	}
+}
+
+function checkFilename(name) {
```

Comment:
There are some valid points here.

In general I'm for simple solutions and not over-engendering things. I would restrict instance naming to readable characters, something like `/[A-Za-z0-9\-\_\. ]+/` - solves a lot of problems, except Windows special file names - those we can add a check for just in case :)

Reply:
And yet your simple solution would let you pass `..` as the instance name, or one that ends with a space, but not a useful name like `Æon`.  I don't see any good reason for why the name should be restricted like that when both Windows and Linux allows any character in the file name besides those listed, and this name is also displayed in the server list, so if your language happen to not be based entirely on pain lain characters then it would be too bad for you.

Names like nul.txt are valid, you can create, open, and delete files and directories named that without problems, not recommended is not the same as not valid.  As for the path length limitation, it cannot be checked for, you simply don't know how deep the directory this program will end up in and I haven't seen any program ever actually checking for it.

I also don't see why this should be outsourced as a library.  It's a 20 line sanity check that the instance directory would actually be a valid directory name (a check that wasn't even performed before), not some kind of malicious user input validator.

Reply:
If you're going that route you might as well replace bad characters with underscore, or use the id as folder name.

File is often used to mean both directory and file so checkFileName is not really wrong, it works on both.

Reply:
So if you name one instance <foo> and another instance "foo" you expect them to be separate?  How is this any different from the existing name collisions that occurs on Windows due to its filesystem being case insensitive?  Foo and foo is the same instance on Windows but not on Linux.

Reply:
Also, I know we have chineese users who write their entire server names with their symbols. Instances named ____________ would be kinda bad

Reply:
By bad characters I mean the 9 reserved ones that Windows does not permit in file names as well as the 32 ASCII control codes.  International characters are not bad characters.